src/server/channel_store.h

```diff
@@ -34,39 +46,108 @@ class ChannelStore {
     void operator=(const Subscriber&) = delete;
 
     // Sort by thread-id. Subscriber without owner comes first.
-    static bool ByThread(const Subscriber& lhs, const Subscriber& rhs) {
-      if (lhs.thread_id == rhs.thread_id)
-        return (lhs.conn_cntx != nullptr) < (rhs.conn_cntx != nullptr);
-      return lhs.thread_id < rhs.thread_id;
-    }
+    static bool ByThread(const Subscriber& lhs, const Subscriber& rhs);
+
+    ConnectionContext* conn_cntx;
+    util::fibers_ext::BlockingCounter borrow_token;  // to keep connection alive
+    uint32_t thread_id;
+    std::string pattern;  // non-empty if registered via psubscribe
   };
 
-  void AddSub(std::string_view channel, ConnectionContext* me, uint32_t thread_id);
-  void RemoveSub(std::string_view channel, ConnectionContext* me);
+  // Centralized controller to prevent overlaping updates.
+  struct ControlBlock {
+    void Destroy();
+
+    ChannelStore* most_recent;
+    ::boost::fibers::mutex update_mu;  // locked during updates.
+  };
```

Comment:
In the future, instead of plain locking, we can accumulate all changes and let all fulfill all writers requests with a single operation

Reply:
Its a bit more difficult to synchronize it, besides we need to explicitly aggregate changes by key to tell whether we'll modify the map slots (replace/add on single value cancel each other)